### PR TITLE
Package errpy.0.0.9

### PR DESCRIPTION
**Pull request by NgaiJustin**
### `errpy.0.0.9`
Errpy: An Error Recovering Python Parser implemented in Rust
This is a new project to provide an error-recovering Python parser, implemented in Rust based on tree-sitter. Our short-term goal is to use it in the pyre type checker, in order to provide better IDE features.

---
* Homepage: https://github.com/facebook/errpy
* Source repo: git+https://github.com/facebook/errpy.git
* Bug tracker: https://github.com/facebook/errpy/issues

---
:camel: Pull-request generated by opam-publish v2.2.0

### CODE REVIEW

**Comment by mseri** [comment]
Many failures are to be expected due to rust's versioning. Is the following also expected?
```
#    Compiling ffi_ocaml v0.0.0 (/home/opam/.opam/4.14/.opam-switch/build/errpy.0.0.9)
# error[E0308]: `match` arms have incompatible types
#     --> printers/ast_pretty_print.rs:1698:49
#      |
# 1691 |                   let res = match rest {
#      |  ___________________________-
# 1692 | |                     ConstantDesc::Str(vala) => vala,
#      | |                                                ---- this is found to be of type `&std::string::String`
# 1693 | |                     ConstantDesc::ByteStr(vala) => {
# 1694 | |                         pprint_output.push_str("b");
# 1695 | |                         vala
#      | |                         ---- this is found to be of type `&std::string::String`
# 1696 | |                     }
# 1697 | |                     ConstantDesc::Num(Num::BigInt(vala)) => vala,
#      | |                                                             ---- this is found to be of type `&std::string::String`
# 1698 | |                     ConstantDesc::Bool(true) => "True",
#      | |                                                 ^^^^^^ expected struct `std::string::String`, found `str`
# ...    |
# 1701 | |                     _ => "",
# 1702 | |                 };
#      | |_________________- `match` arms have incompatible types
#      |
#      = note: expected reference `&std::string::String`
#                 found reference `&'static str`
# 
# For more information about this error, try `rustc --explain E0308`.
# error: could not compile `ffi_ocaml` due to previous error
# warning: build failed, waiting for other jobs to finish...
# error: could not compile `ffi_ocaml` due to previous error
```
If so, do we need to check a different minimal version for rust?

**Comment by NgaiJustin** [comment]
Thanks for looking into this. This seems to be an issue with the old rust versioning. I will try and see if explicitly typing will help patch the issue, otherwise, I can try to specify a minimal version for rust in a new PR

**Comment by NgaiJustin** [comment]
Thanks for looking into this. It seems like this is also a rust versioning error (builds with `rustc 1.70` locally). We already added the dependency on `"conf-rust-2021"`, is there anything else we can add to specify a different minimal version for rust?

**Comment by kit-ty-kate** [comment]
> is there anything else we can add to specify a different minimal version for rust?

what is your actual minimal version for rust? The 2021 edition is too old for you it looks like.

**Comment by NgaiJustin** [comment]
`rustc 1.65` seems to be our minimal version

**Comment by NgaiJustin** [comment]
@kit-ty-kate Just following up, anything we should do for this to be approved?

**Comment by mseri** [comment]
Perhaps we could have a post install (failure) message explaining what is the minimum required rust version. This could help resolve those remaining filires

File: packages/errpy/errpy.0.0.9/opam [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+dev-repo: "git+https://github.com/facebook/errpy.git"
+url {
+  src: "https://github.com/facebook/errpy/archive/refs/tags/0.0.9.tar.gz"
+  checksum: [
+    "md5=ec89b3c48081a43a3b9d8b4216fe2dad"
+    "sha512=a553863a0d80188e2dc52f5cc7b153fbe51f71f8cb991c0c7a524cbe3a000952148150de13daf90032a9723aeb991a666340627502b04ec0f7aeee295f0b86fd"
+  ]
+}
+post-messages: [
+  "ERRPY requires a rustc version of 1.65+ due to the tree-sitter dependency" {rustc:version < "1.65"}
```
Comment (mseri) [inline packages/errpy/errpy.0.0.9/opam:42]:
```suggestion
  "ERRPY requires a rustc version of 1.65+ due to the tree-sitter dependency" {failure}
```
This won't work, opam does not know rustc version. We need to use "failure" here to ensure it is printed each time the package fail

**Comment by NgaiJustin** [comment]
Thanks for the fix @mseri, any estimates for when this will be merged?

File: packages/errpy/errpy.0.0.9/opam [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+synopsis: "Errpy: An Error Recovering Python Parser implemented in Rust"
+description:
+  "This is a new project to provide an error-recovering Python parser, implemented in Rust based on tree-sitter. Our short-term goal is to use it in the pyre type checker, in order to provide better IDE features."
+maintainer: ["pyre@meta.com"]
+authors: ["pyre@meta.com"]
+license: "MIT"
+homepage: "https://github.com/facebook/errpy"
+bug-reports: "https://github.com/facebook/errpy/issues"
+depends: [
+  "ocaml"
```
Comment (mseri) [inline packages/errpy/errpy.0.0.9/opam:12]:
```suggestion
  "ocaml" { >= "4.03"}
```
Due to
```
#=== ERROR while compiling errpy.0.0.9 ========================================#
# context              2.2.0~alpha3~dev | linux/x86_64 | ocaml-base-compiler.4.02.3 | pinned(https://github.com/facebook/errpy/archive/refs/tags/0.0.9.tar.gz)
# path                 ~/.opam/4.02/.opam-switch/build/errpy.0.0.9
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p errpy -j 71 @install
# exit-code            1
# env-file             ~/.opam/log/errpy-7-b519cd.env
# output-file          ~/.opam/log/errpy-7-b519cd.out
### output ###
# File "vendor/ocaml/interop/ocamlrep_ocamlpool/dune", line 1, characters 0-124:
# 1 | (library
# 2 |  (name ocamlpool)
# 3 |  (package errpy)
# 4 |  (wrapped false)
# 5 |  (modules)
# 6 |  (foreign_stubs
# 7 |   (language c)
# 8 |   (names ocamlpool)))
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlopt.opt -w -40 -w A-30 -O3 -a -o vendor/ocaml/interop/ocamlrep_ocamlpool/ocamlpool.cmxa -cclib -locamlpool_stubs)
# /home/opam/.opam/4.02/bin/ocamlopt.opt: unknown option '-O3'.
# Usage: ocamlopt <options> <files>
[…]
# Either upgrade to rustc 1.64.0 or newer, or use
# cargo update -p winnow@0.5.3 --precise ver
# where `ver` is the latest version of `winnow` supporting rustc 1.63.0
# (cd _build/.sandbox/842dcf617d6cf99168011690d140033c/default && .ppx/0bb5094b13567c3c4e1ee33245b80e63/ppx.exe --cookie 'library-name="errpy"' -o ast.pp.ml --impl ast.ml --dump-ast)
# File "ast.ml", line 51, characters 18-19:
# Error: Syntax error
```

**Pull request merged**